src/rust/kj/http.rs

```diff
@@ -0,0 +1,332 @@
+    #[derive(Debug, PartialEq, Eq, Copy, Clone)]
+    #[repr(u32)]
+    pub enum BuiltinIndicesEnum {
+        CONNECTION,
+        KEEP_ALIVE,
+        TE,
+        TRAILER,
+        UPGRADE,
+        CONTENT_LENGTH,
+        TRANSFER_ENCODING,
+        SEC_WEBSOCKET_KEY,
+        SEC_WEBSOCKET_VERSION,
+        SEC_WEBSOCKET_ACCEPT,
+        SEC_WEBSOCKET_EXTENSIONS,
+        HOST,
+        DATE,
+        LOCATION,
+        CONTENT_TYPE,
+        RANGE,
+        CONTENT_RANGE,
+    }
+
+    unsafe extern "C++" {
+        type BuiltinIndicesEnum;
+        type HttpHeaders;
+        fn clone_shallow(this_: &HttpHeaders) -> KjOwn<HttpHeaders>;
+        fn set_header(this_: Pin<&mut HttpHeaders>, id: BuiltinIndicesEnum, value: &str);
+        unsafe fn get_header<'a>(
+            this_: &'a HttpHeaders,
+            id: BuiltinIndicesEnum,
+        ) -> KjMaybe<&'a [u8]>;
+    }
+
+    // --- kj::HttpService ffi
+
+    unsafe extern "C++" {
+        type TlsStarterCallback;
+    }
+
+    /// Corresponds to `kj::HttpConnectSettings`.
+    struct HttpConnectSettings<'a> {
+        use_tls: bool,
+        tls_starter: KjMaybe<Pin<&'a mut TlsStarterCallback>>,
+    }
+
+    unsafe extern "C++" {
+        type AsyncInputStream = crate::io::AsyncInputStream;
+        type AsyncIoStream = crate::io::AsyncIoStream;
+        type ConnectResponse;
+        type HttpServiceResponse;
+        type HttpService;
+
+        /// Corresponds to `kj::HttpService::request`.
+        async fn request(
+            this_: Pin<&mut HttpService>,
+            method: HttpMethod,
+            url: &[u8],
+            headers: &HttpHeaders,
+            request_body: Pin<&mut AsyncInputStream>,
+            response: Pin<&mut HttpServiceResponse>,
+        ) -> Result<()>;
+
+        /// Corresponds to `kj::HttpService::connect`.
+        async fn connect(
+            this_: Pin<&mut HttpService>,
+            host: &[u8],
+            headers: &HttpHeaders,
+            connection: Pin<&mut AsyncIoStream>,
+            response: Pin<&mut ConnectResponse>,
+            settings: HttpConnectSettings<'_>,
+        ) -> Result<()>;
+    }
+
+    // DynHttpService
+    extern "Rust" {
+        type DynHttpService;
+
+        async unsafe fn request<'a>(
+            self: &'a mut DynHttpService,
+            method: HttpMethod,
+            url: &'a [u8],
+            headers: &'a HttpHeaders,
+            request_body: Pin<&'a mut AsyncInputStream>,
+            response: Pin<&'a mut HttpServiceResponse>,
+        ) -> Result<()>;
+
+        async unsafe fn connect<'a>(
+            self: &'a mut DynHttpService,
+            host: &'a [u8],
+            headers: &'a HttpHeaders,
+            connection: Pin<&'a mut AsyncIoStream>,
+            response: Pin<&'a mut ConnectResponse>,
+            settings: HttpConnectSettings<'a>,
+        ) -> Result<()>;
+    }
+
+    impl Box<DynHttpService> {}
```

Comment:
What's this all about?

Reply:
this is how you signal to cxx to generate necessary Box implemenation parts.